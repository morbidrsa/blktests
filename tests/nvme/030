#!/bin/bash
# SPDX-License-Identifier: GPL-2.0+
# Copyright (c) 2019 Johannes Thumshirn
#
# Test resize of a namespace with block device backed ns

. tests/nvme/rc

DESCRIPTION="test resize of a namespace with block device backed ns"
QUICK=1

requires() {
	_have_program nvme && _have_modules loop nvme-loop nvmet && \
		_have_configfs
}

test() {
	echo "Running ${TEST_NAME}"

	modprobe nvmet
	modprobe nvme-loop

	local port
	local nvmedev
	local namespace
	local loop_dev
	local subsys_name="blktests-subsystem-1"

	truncate -s 1G "${TMPDIR}/img"
	loop_dev="$(losetup -f --show "$TMPDIR/img")"

	_create_nvmet_subsystem "${subsys_name}" "${loop_dev}" \
		"91fdba0d-f87b-4c25-b80f-db7be1418b9e"
	port="$(_create_nvmet_port "loop")"
	_add_nvmet_subsys_to_port "${port}" "${subsys_name}"

	nvme connect -t loop -n "${subsys_name}"

	nvmedev="$(_find_nvme_loop_dev)"
	namespace="${nvmedev}n1"
	cat "/sys/block/${namespace}/uuid"
	cat "/sys/block/${namespace}/wwid"

	echo "Check loop device"
	cat /proc/partitions | grep ${loop_dev#/dev/} | awk '{ print $3 }'
	blockdev --getsz ${loop_dev}

	echo "Check NVMe Namespace"
	cat /proc/partitions | grep ${namespace} | awk '{ print $3 }'
	blockdev --getsz /dev/${namespace}

	echo "Resizing"
	truncate -s +1G "${TMPDIR}/img"
	losetup -c ${loop_dev}
	printf 0 > ${NVMET_CFS}/subsystems/${subsys_name}/namespaces/1/enable
	printf 1 > ${NVMET_CFS}/subsystems/${subsys_name}/namespaces/1/enable
	sleep 1

	echo "Check loop device"
	cat /proc/partitions | grep ${loop_dev#/dev/} | awk '{ print $3 }'
	blockdev --getsz ${loop_dev}

	echo "Check NVMe Namespace"
	cat /proc/partitions | grep ${namespace} | awk '{ print $3 }'
	blockdev --getsz /dev/${namespace}

	nvme disconnect -n "${subsys_name}"

	_remove_nvmet_subsystem_from_port "${port}" "${subsys_name}"
	_remove_nvmet_subsystem "${subsys_name}"
	_remove_nvmet_port "${port}"

	losetup -d "$loop_dev"

	rm "$TMPDIR/img"

	modprobe -r nvme-loop
	modprobe -r nvmet

	echo "Test complete"
}
