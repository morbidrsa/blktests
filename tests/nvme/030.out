Running nvme/030
91fdba0d-f87b-4c25-b80f-db7be1418b9e
uuid.91fdba0d-f87b-4c25-b80f-db7be1418b9e
Check loop device
1048576
2097152
Check NVMe Namespace
1048576
2097152
Resizing
Check loop device
2097152
4194304
Check NVMe Namespace
2097152
4194304
NQN:blktests-subsystem-1 disconnected 1 controller(s)
Test complete
